### PR TITLE
test(fiber): Pool tests

### PR DESCRIPTION
**Pull request by rgrinberg**
Demonstrates that we incorrectly allow pools to run more than once

### CODE REVIEW

**Review by snowleopard — COMMENTED** [review]
Can you add some comments? I don't really understand what these tests do.

**Comment by rgrinberg** [comment]
Added some descriptions to the tests

File: test/expect-tests/fiber/fiber_tests.ml [hunk 1]
```diff
@@ -847,6 +847,60 @@ let%expect_test "raise exception" =
...
+  (Scheduler.run
+  @@
+  let pool = Pool.create () in
+  let* () = Pool.stop pool in
+  let* () = Pool.run pool in
+  Pool.run pool);
+  [%expect.unreachable]
+  [@@expect.uncaught_exn {| (Test_scheduler.Never) |}]
+
+let%expect_test "pool - weird deadlock" =
```
Comment (rgrinberg) [inline test/expect-tests/fiber/fiber_tests.ml:884]: @snowleopard this one wasted a lot of my time when implementing the action runner. Let's see if you know why this happens offhand :)

File: test/expect-tests/fiber/fiber_tests.ml [hunk 2]
```diff
@@ -847,6 +847,60 @@ let%expect_test "raise exception" =
        (fun () -> Pool.stop pool));
   [%expect {| Caught Exit |}]
 
+let%expect_test "double run a pool" =
+  (* Calling [Pool.run] on the same pool shouldn't be allowed *)
```
Comment (snowleopard) [inline test/expect-tests/fiber/fiber_tests.ml:851]: Could you explain why it shouldn't be allowed? (I don't know much about the pools, sorry.)

File: test/expect-tests/fiber/fiber_tests.ml (hunk 1 above)
Comment (snowleopard) [reply test/expect-tests/fiber/fiber_tests.ml:884]: I have no clue what's happening here, so no guesses from me!

File: test/expect-tests/fiber/fiber_tests.ml (hunk 2 above)
Comment (rgrinberg) [reply test/expect-tests/fiber/fiber_tests.ml:851]: Currently, `Pool.run` means to run all the tasks inside the pool until a `stop` has been given.

So if we run two `Pool.run`, they will compete for the same queue of tasks and it will be unclear which `run` will get which task. This is really important when tasks raise exceptions as we want to know exactly where uncaught task exceptions should be caught (around `run`).

File: test/expect-tests/fiber/fiber_tests.ml (hunk 1 above)
Comment (rgrinberg) [reply test/expect-tests/fiber/fiber_tests.ml:884]: The issue is the underlying mvar implementation. Writing to an mvar that is empty does not block, so the first test happily succeeds. The second test fails because the second `Pool.task` blocks on writing to the mvar.

All of this makes sense when dealing with mvars - but pools aren't mvars. Pools are just used to "reap" sets of fibers, that is wait for them to complete and handle any of their exceptions if need be. So it doesn't make sense for them to inherit this behavior at all.

File: test/expect-tests/fiber/fiber_tests.ml (hunk 2 above)
Comment (snowleopard) [reply test/expect-tests/fiber/fiber_tests.ml:851]: Great! Please copy this text as a comment.

File: test/expect-tests/fiber/fiber_tests.ml [hunk 3]
```diff
@@ -847,6 +847,60 @@ let%expect_test "raise exception" =
...
+  (Scheduler.run
+  @@
+  let pool = Pool.create () in
+  Fiber.fork_and_join_unit (fun () -> Pool.run pool) (fun () -> Pool.run pool));
+  [%expect.unreachable]
+  [@@expect.uncaught_exn {| (Test_scheduler.Never) |}]
+
+let%expect_test "start -> stop -> start a pool" =
+  (* We shouldn't be able to call [Pool.run] after stopping it *)
```
Comment (snowleopard) [inline test/expect-tests/fiber/fiber_tests.ml:860]: Why? According to the `Pool` docs, `stop` only affects subsequent `task` calls.

File: test/expect-tests/fiber/fiber_tests.ml [hunk 4]
```diff
@@ -847,6 +847,60 @@ let%expect_test "raise exception" =
...
+  (Scheduler.run
+  @@
+  let pool = Pool.create () in
+  let* () =
+    Fiber.fork_and_join_unit
+      (fun () -> Pool.run pool)
+      (fun () -> Pool.stop pool)
+  in
+  Pool.run pool);
```
Comment (snowleopard) [inline test/expect-tests/fiber/fiber_tests.ml:869]: Isn't this actually another case of double run?

File: test/expect-tests/fiber/fiber_tests.ml [hunk 5]
```diff
@@ -847,6 +847,60 @@ let%expect_test "raise exception" =
...
+  [@@expect.uncaught_exn {| (Test_scheduler.Never) |}]
+
+let%expect_test "stop a pool and then run it" =
+  (* Weird deadlock when we call [stop] after [run] *)
+  (Scheduler.run
+  @@
+  let pool = Pool.create () in
+  let* () = Pool.stop pool in
+  let* () = Pool.run pool in
+  Pool.run pool);
```
Comment (snowleopard) [inline test/expect-tests/fiber/fiber_tests.ml:880]: I don't understand: you again have a double `run` here, not a `stop` after `run` as the comment says.

File: test/expect-tests/fiber/fiber_tests.ml (hunk 3 above)
Comment (rgrinberg) [reply test/expect-tests/fiber/fiber_tests.ml:860]: Clarified it. Here we want to forbid calling `run`, then `stop`, then `run`. That is, one cannot "reuse" a pool even if with start/stop. There's no reason why this shouldn't be allowed, but we just don't support it at the moment.

File: test/expect-tests/fiber/fiber_tests.ml (hunk 4 above)
Comment (rgrinberg) [reply test/expect-tests/fiber/fiber_tests.ml:869]: Yes, but here it's interleaved by a stop. Will add a comment.

File: test/expect-tests/fiber/fiber_tests.ml (hunk 5 above)
Comment (rgrinberg) [reply test/expect-tests/fiber/fiber_tests.ml:880]: Ah yup, the 2nd run isn't necessary. The bug just needs stop and then run. Fixed.

File: test/expect-tests/fiber/fiber_tests.ml [hunk 6]
```diff
@@ -847,6 +847,63 @@ let%expect_test "raise exception" =
        (fun () -> Pool.stop pool));
   [%expect {| Caught Exit |}]
 
+let%expect_test "double run a pool" =
+  (* Calling [Pool.run] on the same pool shouldn't be allowed
```
Comment (snowleopard) [inline test/expect-tests/fiber/fiber_tests.ml:851]:
```suggestion
  (* Calling [Pool.run] twice on the same pool shouldn't be allowed
```

File: test/expect-tests/fiber/fiber_tests.ml [hunk 7]
```diff
@@ -847,6 +847,63 @@ let%expect_test "raise exception" =
...
+
+     We can't allow competing [Pool.run] calls to get tasks. We want to make
+     sure only a [run] will get the exception from all tasks *)
```
Comment (snowleopard) [inline test/expect-tests/fiber/fiber_tests.ml:854]:
```suggestion
     sure only a single [run] will get the exceptions from all the tasks in the pool *)
```

File: test/expect-tests/fiber/fiber_tests.ml [hunk 8]
```diff
@@ -847,6 +847,63 @@ let%expect_test "raise exception" =
...
+  (Scheduler.run
+  @@
+  let pool = Pool.create () in
+  Fiber.fork_and_join_unit (fun () -> Pool.run pool) (fun () -> Pool.run pool));
+  [%expect.unreachable]
+  [@@expect.uncaught_exn {| (Test_scheduler.Never) |}]
+
+let%expect_test "start -> stop -> start a pool" =
```
Comment (snowleopard) [inline test/expect-tests/fiber/fiber_tests.ml:862]:
```suggestion
let%expect_test "run -> stop -> run a pool" =
```

File: test/expect-tests/fiber/fiber_tests.ml [hunk 9]
```diff
@@ -847,6 +847,63 @@ let%expect_test "raise exception" =
...
+    Fiber.fork_and_join_unit
+      (fun () -> Pool.run pool)
+      (fun () -> Fiber.Pool.task pool ~f:(fun () -> Pool.stop pool))
+  in
+  Pool.run pool);
+  [%expect.unreachable]
+  [@@expect.uncaught_exn {| (Test_scheduler.Never) |}]
+
+let%expect_test "stop a pool and then run it" =
+  (* Weird deadlock when we call [stop] after [run] *)
```
Comment (snowleopard) [inline test/expect-tests/fiber/fiber_tests.ml:878]:
```suggestion
  (* Weird deadlock when we call [run] after [stop] *)
```

**Review by snowleopard — APPROVED** [review]
Thanks, I think the tests make sense now. I left a couple of further suggestions.

**Comment by rgrinberg** [comment]
Waiting for the coverage to complete before merging. I would like to see what edge cases are left untested.
